:seedling: PartitionSet e2e

fgiloux opened the pull request:
## Summary

This PR adds end-to-end tests for PartitionSet reconciliation.

## Related issue(s)

Contributes to #2334

fgiloux commented:
/hold
It will need to be rebased after #2513 has merged

fgiloux commented:
/retest

p0lyn0mial commented on pkg/reconciler/topology/partitionset/resources.go:
```diff
@@ -35,6 +36,7 @@ func generatePartition(name string, matchExpressions []metav1.LabelSelectorRequi
 	for _, label := range labels {
 		pname = pname + "-" + strings.ToLower(matchLabels[label])
 	}
+	pname = pname[:min(validation.DNS1123SubdomainMaxLength-1, len(pname))]
```
so, there is a limit :)

p0lyn0mial replied:
please add a unit test

stevekuznetsov replied:
What about collisions? This is why we use the hash in many other places.

fgiloux replied:
GenerateName is used to avoid collisions.

p0lyn0mial commented on pkg/reconciler/topology/partitionset/partitionset_reconcile.go:
```diff
@@ -89,10 +89,11 @@ func (c *controller) reconcile(ctx context.Context, partitionSet *topologyv1alph
 	}
 
 	var matchLabelsMap map[string]map[string]string
+	dimensions := removeDuplicates(partitionSet.Spec.Dimensions)
```
should we put it into validation/cel ?
that way we wouldn't get any duplicates, right ?

p0lyn0mial replied:
@sttts wdyt?

p0lyn0mial commented on test/e2e/reconciler/partition/partitionset_test.go:
```diff
@@ -0,0 +1,332 @@
+limitations under the License.
+*/
+
+package partition
```
should we rename `test/e2e/reconciler/partition/partitionset_test.go` to `test/e2e/reconciler/partitionset/partitionset_test.go` ?

p0lyn0mial commented on test/e2e/reconciler/partition/partitionset_test.go:
```diff
@@ -0,0 +1,332 @@
+		}
+		return false, spew.Sdump(partitionSet.Status.Conditions)
+	}, wait.ForeverTestTimeout, 100*time.Millisecond, "expected valid partitionSet")
+
```
should we check that no `Partition` was created ?

stevekuznetsov replied:
IMHO we should focus a bit more on the happy cases in e2e, exhaustive negative checking can be fragile as we've learned and requires private fixtures...

p0lyn0mial commented on test/e2e/reconciler/partition/partitionset_test.go:
```diff
@@ -0,0 +1,332 @@
+		partitions, err = partitionClient.Cluster(partitionClusterPath).List(ctx, metav1.ListOptions{})
+		require.NoError(t, err, "error retrieving partitions")
+		if len(partitions.Items) == 1 {
+			return true, ""
```
should we examine `Partition.Spec.Selector` ?

p0lyn0mial commented on test/e2e/reconciler/partition/partitionset_test.go:
```diff
@@ -0,0 +1,332 @@
+		partitions, err = partitionClient.Cluster(partitionClusterPath).List(ctx, metav1.ListOptions{})
+		require.NoError(t, err, "error retrieving partitions")
+		if len(partitions.Items) == 2 {
+			return true, ""
```
should we check `Partition.Spec.Selector` ?

stevekuznetsov replied:
We need to validate that we get the right ones ehre

p0lyn0mial commented on test/e2e/reconciler/partition/partitionset_test.go:
```diff
@@ -0,0 +1,332 @@
+		return false, fmt.Sprintf("expected 1 partition, but got %d", len(partitions.Items))
+	}, wait.ForeverTestTimeout, 100*time.Millisecond, "expected 1 partition")
+
+	// The following tests are focused on the admission
```
I'd move it to a new separate test that simply checks admission

fgiloux replied:
I thought of that but did not do it due to the cost of running a private cluster.

p0lyn0mial commented on test/e2e/reconciler/partition/partitionset_test.go:
```diff
@@ -0,0 +1,332 @@
+	}
+	_, err = partitionSetClient.Cluster(partitionClusterPath).Create(ctx, errorPartitionSet, metav1.CreateOptions{})
+	require.Error(t, err, "error creating partitionSet expected")
+
```
should we check that a `PartitionSet` wasn't created ?

p0lyn0mial commented on test/e2e/reconciler/partition/partitionset_test.go:
```diff
@@ -0,0 +1,332 @@
+	_, err = partitionSetClient.Cluster(partitionClusterPath).Create(ctx, errorPartitionSet, metav1.CreateOptions{})
+	require.Error(t, err, "error creating partitionSet expected")
+
+	t.Logf("Character not allowed at first place in matchExpressions values")
```
Not sure if we need to check all incorrect values, i'd rely on k8s's validation.
From our perspective the most important thing is to make sure an invalid request will be rejected and won't have any side effect (i.e. deleting a partition).

p0lyn0mial commented on test/e2e/reconciler/partition/partitionset_test.go:
```diff
@@ -0,0 +1,332 @@
+	_, err = partitionSetClient.Cluster(partitionClusterPath).Create(ctx, errorPartitionSet, metav1.CreateOptions{})
+	require.Error(t, err, "error creating partitionSet expected")
+
+	t.Logf("Partition name cut when the label values sum up")
```
I'm okay with checking it on a unit test level.

stevekuznetsov commented on pkg/reconciler/topology/partitionset/partitionset_reconcile.go:
```diff
@@ -215,3 +216,16 @@ func partition(shards []*corev1alpha1.Shard, dimensions []string, shardSelectorL
 }
+
+// removeDuplicates makes sure that a value is not found more than once in the slice.
+func removeDuplicates(slice []string) []string {
```
```go
sets.NewString(slice).List()
```

fgiloux replied:
changed

stevekuznetsov commented on pkg/reconciler/topology/partitionset/resources.go:
```diff
@@ -48,3 +50,10 @@ func generatePartition(name string, matchExpressions []metav1.LabelSelectorRequi
 	}
 }
+
+func min(a, b int) int {
```
IMHO we don't need to accumulate this sort of logic. Just inline it.

fgiloux replied:
inlining is done by the compiler. I find this way more readable and it seems that it is not an uncommon practice.

stevekuznetsov commented on test/e2e/reconciler/partition/partitionset_test.go:
```diff
@@ -0,0 +1,332 @@
+	t.Parallel()
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	server := framework.PrivateKcpServer(t)
```
Why?

stevekuznetsov commented on test/e2e/reconciler/partition/partitionset_test.go:
```diff
@@ -0,0 +1,332 @@
+	defer cancel()
+	server := framework.PrivateKcpServer(t)
+
+	// Create organization and workspace
```
Do we need an org anymore? Why?

fgiloux replied:
we don't. It does not hurt either

stevekuznetsov commented on test/e2e/reconciler/partition/partitionset_test.go:
```diff
@@ -0,0 +1,332 @@
+	}
+	partitionSet, err = partitionSetClient.Cluster(partitionClusterPath).Create(ctx, partitionSet, metav1.CreateOptions{})
+	require.NoError(t, err, "error creating partitionSet")
+	framework.Eventually(t, func() (bool, string) {
```
```
framework.EventuallyCondition
```

stevekuznetsov commented on test/e2e/reconciler/partition/partitionset_test.go:
```diff
@@ -0,0 +1,332 @@
+	framework.Eventually(t, func() (bool, string) {
+		partitionSet, err = partitionSetClient.Cluster(partitionClusterPath).Get(ctx, partitionSet.Name, metav1.GetOptions{})
+		require.NoError(t, err, "error retrieving partitionSet")
+		if conditions.IsTrue(partitionSet, topologyv1alpha1.PartitionSetValid) && conditions.IsTrue(partitionSet, topologyv1alpha1.PartitionsReady) {
```
Roll this up in the reconciler to a ready condition please, waiting on more than one is fragile

fgiloux replied:
this mechanism is tested here.

stevekuznetsov commented on test/e2e/reconciler/partition/partitionset_test.go:
```diff
@@ -0,0 +1,332 @@
+	shardClient := kcpClusterClient.CoreV1alpha1().Shards()
+	_, err = shardClient.Cluster(core.RootCluster.Path()).Create(ctx, shard1a, metav1.CreateOptions{})
+	require.NoError(t, err, "error creating shard")
+	framework.Eventually(t, func() (bool, string) {
```
```
framework.EventuallyCondition
```

fgiloux replied:
did not exist when this test was written, does not allow to check values of multiple conditions.

stevekuznetsov commented on test/e2e/reconciler/partition/partitionset_test.go:
```diff
@@ -0,0 +1,332 @@
+		return false, fmt.Sprintf("expected 1 partition, but got %d", partitionSet.Status.Count)
+	}, wait.ForeverTestTimeout, 100*time.Millisecond, "expected the partition count to be 1")
+	partitionClient := kcpClusterClient.TopologyV1alpha1().Partitions()
+	var partitions *topologyv1alpha1.PartitionList
```
Please continue to `t.Logf()` the steps

stevekuznetsov commented on test/e2e/reconciler/partition/partitionset_test.go:
```diff
@@ -0,0 +1,332 @@
+		return false, fmt.Sprintf("expected 1 partition, but got %d", len(partitions.Items))
+	}, wait.ForeverTestTimeout, 100*time.Millisecond, "expected 1 partition")
+
+	t.Logf("Creating a shard part of a second region")
```
```suggestion
	t.Logf("Creating a shard in a second region")
```

stevekuznetsov commented on test/e2e/reconciler/partition/partitionset_test.go:
```diff
@@ -0,0 +1,332 @@
+	}
+	shard2, err = shardClient.Cluster(core.RootCluster.Path()).Create(ctx, shard2, metav1.CreateOptions{})
+	require.NoError(t, err, "error creating shard")
+	framework.Eventually(t, func() (bool, string) {
```
eventuallycondition

stevekuznetsov reviewed (commented):
IMHO we should write these tests on the shared fixture. Generate your labels and shard names with random suffixes to allow these tests to co-exist even with other instances of the same test in parallel, and mark the shards you're creating as not ready for scehduling so as to not impact any other part of the system.

fgiloux commented:
> IMHO we should write these tests on the shared fixture. Generate your labels and shard names with random suffixes to allow these tests to co-exist even with other instances of the same test in parallel, and mark the shards you're creating as not ready for scehduling so as to not impact any other part of the system.

I had proposed [something similar in a past PR](https://github.com/kcp-dev/kcp/pull/2608#discussion_r1072187994). I see you have [relaunched a discussion in slack](https://kubernetes.slack.com/archives/C021U8WSAFK/p1676036604442249) on that. The change in the approach is out of scope of this PR.

stevekuznetsov commented:
/hold

fgiloux commented:
> Sorry, I don't think we need to check in more tests using private fixtures. We're already at critical mass with e2e flakiness and adding much more load there is not reasonable. How hard is it to implement some "not-ready-for-scheduling" field on the shard? We're in alpha versions, can iterate.

Great! Give me a shout when you agree on something.

stevekuznetsov commented:
Sounds like a field `shard.spec.schedulable={true,false}` would suffice, and we'd just need to update the workspace reconciler to choose only the schedulable subset when `workspace.spec.location` is unset.

p0lyn0mial commented:
FYI https://github.com/kcp-dev/kcp/pull/2782 should unblock this PR.

fgiloux commented:
> FYI #2782 should unblock this PR.

This does not seem to be sufficient:
~~~
    apibinding_test.go:348: === Testing identity wildcards
    apibinding_test.go:379: Verify "root:e2e-workspace-wqp4k:consumer-1-bound-against-1" bound to service provider 1 ("root:e2e-workspace-wqp4k:service-provider-1") wildcard list works
    apibinding_test.go:351: Get APIBinding for workspace root:e2e-workspace-wqp4k:consumer-1-bound-against-1
    apibinding_test.go:360: Doing a wildcard identity list for wildwest.dev/v1alpha1, Resource=cowboys:6440e16b668082bc273a03875cca56af91f1855ab12089d0a1d79dfdd1e8019b against root:e2e-workspace-wqp4k:consumer-1-bound-against-1 workspace on shard partition-shard-1a
    apibinding_test.go:361: kubeconfig for shard "partition-shard-1a" not found
~~~

fgiloux commented:
/unhold

fgiloux commented:
@p0lyn0mial @stevekuznetsov I have refactored the tests to use a shared server. I have also added a commit to patch some sensitivity to non-schedulable shards within the APIBinding tests.
Can you please have a look? I am hoping we can get this PR merged now.

sttts commented:
/lgtm
/approve
